framework/src/executioners/Transient.C

```diff
@@ -228,33 +228,36 @@ Transient::postStep()
 void
 Transient::execute()
 {
-
   preExecute();
 
+  if (_app.isRecovering())
+    incrementStepOrReject();
+
   // NOTE: if you remove this line, you will see a subset of tests failing. Those tests might have a wrong answer and might need to be regolded.
   // The reason is that we actually move the solution back in time before we actually start solving (which I think is wrong).  So this call here
   // is to maintain backward compatibility and so that MOOSE is giving the same answer.  However, we might remove this call and regold the test
   // in the future eventually.
-  if (!_app.isRecovering())
+  else
     _problem.advanceState();
 
+
   // Start time loop...
-  while (true)
+  while (keepGoing())
   {
-    if (_first != true)
-      incrementStepOrReject();
-
+#ifdef LIBMESH_ENABLE_AMR
+    if (!_first && _problem.adaptivity().isOn())
+      _problem.adaptMesh();
+#endif
```

Comment:
I'm looking at this again and @andrsd's proposal is actually pretty close to the way it was before with the exception of the fix. The loop itself is less clean that way (we're back to a `while(true)` loop with a break in the middle) but I suppose it solves the time problem.